optuna/_transform.py

```diff
@@ -0,0 +1,271 @@
+            corresponding distributions in log space during transformation/untransformation.
+            Should always be :obj:`True` if any parameters are going to be sampled from the
+            transformed space.
+        transform_step:
+            If :obj:`True`, offset the lower and higher bounds by a half step each, increasing the
+            space by one step. This allows fair sampling for values close to the bounds.
+            Should always be :obj:`True` if any parameters are going to be sampled from the
+            transformed space.
+
+    Attributes:
+        bounds:
+            Constructed bounds from the given search space.
+
+    Note:
+        Parameter values are not scaled to the unit cube.
+
+    Note:
+        ``transform_log`` and ``transform_step`` are useful for constructing bounds and parameters
+        without any actual transformations by setting those arguments to :obj:`False`. This is
+        needed for e.g. the hyperparameter importance assessments.
+
+    """
+
+    def __init__(
+        self,
+        search_space: Dict[str, BaseDistribution],
+        transform_log: bool = True,
+        transform_step: bool = True,
+    ) -> None:
+        # TODO(hvy): Avoid casting to `OrderedDict` when Python 3.6 is no longer supported since
+        # order will be guaranteed.
+        search_space = OrderedDict(search_space)
+
+        bounds, column_to_encoded_columns = _transform_search_space(
+            search_space, transform_log, transform_step
+        )
+
+        self._bounds = bounds
+        self._column_to_encoded_columns = column_to_encoded_columns
+        self._search_space = search_space
+        self._transform_log = transform_log
+
+    @property
+    def bounds(self) -> numpy.ndarray:
+        return self._bounds
+
+    def transform(self, params: Dict[str, Any]) -> numpy.ndarray:
+        """Transform a parameter configuration from actual values to continuous space.
+
+        Args:
+            params:
+                A parameter configuration to transform.
+
+        Returns:
+            An 1-dimenstional ``numpy.ndarray`` holding the transformed parameters in the
+            configuration.
+
+        """
+        trans_params = numpy.zeros(self._bounds.shape[0], dtype=numpy.float64)
+
+        bound_idx = 0
+        for name, distribution in self._search_space.items():
+            assert name in params, "Parameter configuration must contain all distributions."
+            param = params[name]
+
+            if isinstance(distribution, CategoricalDistribution):
+                choice_idx = distribution.to_internal_repr(param)
+                trans_params[bound_idx + choice_idx] = 1
+                bound_idx += len(distribution.choices)
+            else:
+                trans_params[bound_idx] = _transform_numerical_param(
+                    param, distribution, self._transform_log
+                )
```

Comment:
Currently, the transformation logic of the search space and the parameter value is different. I'm not sure but I think we should unify them. However, the current logic of those transformations is different in `CmaEsSampler`. @hvy @c-bata What do you think?

Reply:
I noticed that the transformation of the search space of this PR is used in `CmaEsSampler`.

Reply:
> the transformation logic of the search space and the parameter value is different. I'm not sure but I think we should unify them.

Sorry, I couldn't understand exactly in what way you meant that they are different. Could you explain how we could do the transformation instead?

Reply:
Sorry for the confusion. I was wrong about the difference. Let me resolve this conversation.